File: X1/Spam.java
import java.util.Scanner;

public class Spam
{
	public static void main(String[] args)
	{
		int space = 0;
		int lower = 0;
		int upper = 0;
		int symbols = 0;

		String in = new Scanner(System.in).nextLine();
		char[] chars = in.toCharArray();
		double length = chars.length;

		for (char c : chars) {
			int code = (int) c;
			if (c == '_')
				space++;
			else if (code >= 65 && code <= 90)
				upper++;
			else if (code >= 97 && code <= 122)
				lower++;
			else
				symbols++;
		}

		System.out.println(space / length);
		System.out.println(lower / length);
		System.out.println(upper / length);
		System.out.println(symbols / length);
	}
}

File: X1/spam.hs
module Main where

import Data.Char

-- Helper function to divide integers without Integral issues
(//) :: Int -> Int -> Double
(//) a b = fromIntegral (a) / fromIntegral (b)

main = do
    input <- getLine
    let space = length $ filter (== '_') input
    let lower = length $ filter (isLower) input
    let upper = length $ filter (isUpper) input
    let symbols = length input - (space + lower + upper)
    print $ space // length input
    print $ lower // length input
    print $ upper // length input
    print $ symbols // length input